jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/profiles/ProfileCredentialProviderFactory.kt

```diff
@@ -3,26 +3,293 @@
+
+private class ProfileCredentialsIdentifier(internal val profileName: String) : ToolkitCredentialsIdentifier() {
+    override val id = "profile:$profileName"
+    override val displayName = message("credentials.profile.name", profileName)
+    override val factoryId = PROFILE_FACTORY_ID
+}
+
+class ProfileCredentialProviderFactory : CredentialProviderFactory, Disposable {
+
+    private val profileWatcher = ProfileWatcher(this)
+    private val profileHolder = ProfileHolder()
+
+    override val id = PROFILE_FACTORY_ID
+
+    override fun setUp(credentialLoadCallback: CredentialsChangeListener) {
+        // Load the initial data, then start the background watcher
+        loadProfiles(credentialLoadCallback, true)
 
-    companion object {
-        val profileWatcher = ProfileWatcher().also {
-            it.start()
-            // TODO: Scope this better in the cred manager refactor
-            Disposer.register(ApplicationManager.getApplication(), it)
+        profileWatcher.start {
+            loadProfiles(credentialLoadCallback, false)
         }
     }
+
+    private fun loadProfiles(credentialLoadCallback: CredentialsChangeListener, initialLoad: Boolean) {
+        val profilesAdded = mutableListOf<ProfileCredentialsIdentifier>()
+        val profilesModified = mutableListOf<ProfileCredentialsIdentifier>()
+        val profilesRemoved = mutableListOf<ProfileCredentialsIdentifier>()
+
+        val previousProfilesSnapshot = profileHolder.snapshot()
+        val newProfiles = try {
+            validateAndGetProfiles()
+        } catch (e: Exception) {
+            notifyUserOfLoadFailure(e)
+
+            TelemetryService.recordSimpleTelemetry(null, "aws_credentials_load", false)
+
+            return
+        }
+
+        newProfiles.validProfiles.forEach {
+            val previousProfile = previousProfilesSnapshot.remove(it.key)
+            if (previousProfile == null) {
+                // It was not in the snapshot, so it must be new
+                profilesAdded.add(ProfileCredentialsIdentifier(it.key))
+            } else {
+                // If the profile was modified, notify people, else do nothing
+                if (previousProfile != it.value) {
+                    profilesModified.add(ProfileCredentialsIdentifier(it.key))
+                }
+            }
+        }
+
+        // Any remaining profiles must have either become invalid or removed from the cred/config files
+        previousProfilesSnapshot.keys.asSequence().map { ProfileCredentialsIdentifier(it) }.toCollection(profilesRemoved)
+
+        profileHolder.update(newProfiles.validProfiles)
+        credentialLoadCallback(CredentialIdentifierChange(profilesAdded, profilesModified, profilesRemoved))
+
+        notifyUserOfResult(newProfiles, initialLoad)
+    }
+
+    private fun notifyUserOfLoadFailure(e: Exception) {
+        val loadingFailureMessage = message("credentials.profile.failed_load")
+
+        val detail = e.message?.let {
```

Comment:
nit: why is this not part of `credentials.profile.failed_load`?

Reply:
Its so we don't render a ':' if we have nothing after it.

I think there is a way to do that through localization though

Reply:
We could append it in the message but that isn't much better, seems fine as is